frame/staking/src/tests.rs

```diff
@@ -77,2101 +77,2184 @@ macro_rules! gen_paired_account {
+		bond_nominator(8, StakingBalance::KtonBalance(1), vec![3, 5]);
+
+		start_era(1);
+
+		assert_eq_uvec!(validator_controllers(), vec![4, 2]);
+
+		// This test will fail this. Will saturate.
+		check_exposure_all();
+		assert_eq!(Staking::stakers(3).total_power, TOTAL_POWER / 2);
+		assert_eq!(Staking::stakers(5).total_power, TOTAL_POWER / 2);
+	})
+}
+
+// TODO: overflow, should be checked
+// thread 'tests::phragmen_should_not_overflow_nominators' panicked at 'attempt to add with overflow'
+#[test]
+fn phragmen_should_not_overflow_nominators() {
+	ExtBuilder::default().nominate(false).build().execute_with(|| {
+		let _ = Staking::chill(Origin::signed(10));
+		let _ = Staking::chill(Origin::signed(20));
+
+		bond_validator(2, StakingBalance::RingBalance(1));
+		bond_validator(4, StakingBalance::KtonBalance(1));
+
+		bond_nominator(6, StakingBalance::RingBalance(CAP), vec![3, 5]);
+		bond_nominator(8, StakingBalance::KtonBalance(CAP), vec![3, 5]);
+
+		start_era(1);
+
+		assert_eq_uvec!(validator_controllers(), vec![4, 2]);
+
+		// Saturate.
+		assert_eq!(Staking::stakers(3).total_power, TOTAL_POWER / 2);
+		assert_eq!(Staking::stakers(5).total_power, TOTAL_POWER / 2);
+	})
+}
+
+// TODO: check why overflow here
+// thread 'tests::phragmen_should_not_overflow_ultimate' panicked at 'attempt to add with overflow',
+#[test]
+fn phragmen_should_not_overflow_ultimate() {
+	ExtBuilder::default().nominate(false).build().execute_with(|| {
+		bond_validator(2, StakingBalance::RingBalance(CAP));
+		bond_validator(4, StakingBalance::KtonBalance(CAP));
+
+		bond_nominator(6, StakingBalance::RingBalance(CAP), vec![3, 5]);
+		bond_nominator(8, StakingBalance::KtonBalance(CAP), vec![3, 5]);
+
+		start_era(1);
+
+		assert_eq_uvec!(validator_controllers(), vec![4, 2]);
+
+		// Saturate.
+		assert_eq!(Staking::stakers(3).total_power, TOTAL_POWER / 2 - TOTAL_POWER / 20);
+		assert_eq!(Staking::stakers(5).total_power, TOTAL_POWER / 2 + TOTAL_POWER / 20);
+	})
+}
+
+// TODO: check why overflow here
+//thread 'tests::reward_validator_slashing_validator_doesnt_overflow' panicked at 'assertion failed: `(left == right)`
+//left: `8589934590`,
+//right: `17179869180`', frame/staking/src/tests.rs:1936:9
+#[test]
+fn reward_validator_slashing_validator_doesnt_overflow() {
+	ExtBuilder::default().build().execute_with(|| {
+		let stake = CAP * 2;
+		let reward_slash = CAP * 2;
+
+		// Assert multiplication overflows in balance arithmetic.
+		assert!(stake.checked_mul(reward_slash).is_none());
+
+		// Set staker
+		let _ = Ring::make_free_balance_be(&11, stake);
+		<Stakers<Test>>::insert(
+			&11,
+			Exposure {
+				own_ring_balance: stake,
+				total_power: stake as u32,
+				own_kton_balance: 0,
+				own_power: 0,
+				others: vec![],
+			},
+		);
+
+		// Check reward
+		// @required: REVIEW
+		let _ = Staking::reward_validator(&11, reward_slash);
+		assert_eq!(Ring::total_balance(&11), stake);
+
+		// Set staker
```

Comment:
The problem is here, if the own_power above is 0, `ring_balance` in account 11 will not change after `reward`, this make me confused, and it will be `stake * 2` as substrate's tests, only if `own_power = stake`.

Reply:
I guess our testing Exposure should different with Substrate's, refer

https://github.com/darwinia-network/darwinia/pull/331/files/a8a5722f7c5fc16691d77f914b65b877b8525676#r390468901